File: CartaoFidelidade/Services/IbgeService.swift
//
//  IbgeService.swift
//  CartaoFidelidade
//
//  Serviço para consulta à API do IBGE (estados e municípios).
//  100% gratuito, sem API key.
//

import Foundation

struct IbgeState: Decodable {
    let id: Int
    let nome: String
    let sigla: String
}

struct IbgeCity: Decodable {
    let id: Int
    let nome: String
}

enum IbgeService {
    private static let base = "https://servicodados.ibge.gov.br/api/v1/localidades"
    
    static func fetchStates() async throws -> [IbgeState] {
        let url = URL(string: "\(base)/estados?orderBy=nome")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode([IbgeState].self, from: data)
    }
    
    static func fetchCities(stateCode: String) async throws -> [IbgeCity] {
        let url = URL(string: "\(base)/estados/\(stateCode)/municipios?orderBy=nome")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode([IbgeCity].self, from: data)
    }
}

File: src/services/ibgeService.ts
/**
 * Serviço para consulta à API do IBGE (estados e municípios).
 * 100% gratuito, sem API key.
 */

export interface IbgeState {
  id: number;
  nome: string;
  sigla: string;
}

export interface IbgeCity {
  id: number;
  nome: string;
}

const IBGE_BASE = "https://servicodados.ibge.gov.br/api/v1/localidades";

export async function fetchStates(): Promise<IbgeState[]> {
  const res = await fetch(`${IBGE_BASE}/estados?orderBy=nome`);
  const data = await res.json();
  return data;
}

export async function fetchCitiesByState(stateCode: string): Promise<IbgeCity[]> {
  const res = await fetch(`${IBGE_BASE}/estados/${stateCode}/municipios?orderBy=nome`);
  const data = await res.json();
  return data;
}
